programs/svm-spoke/src/state/refund_account.rs

```diff
@@ -0,0 +1,132 @@
+// token account or a claim account. This enum is used to differentiate between the two types.
+pub enum RefundAccount<'info> {
+    TokenAccount(InterfaceAccount<'info, TokenAccount>),
+    ClaimAccount(Account<'info, ClaimAccount>),
+}
+
+impl<'c, 'info> RefundAccount<'info>
+where
+    'c: 'info,
+{
+    // This function is used to parse a refund account from the remaining accounts list. It first tries to parse it as
+    // a token account and if that fails, it tries to parse it as a claim account.
+    pub fn try_from_remaining_account(
+        remaining_accounts: &'c [AccountInfo<'info>],
+        index: usize,
+        expected_token_account: &Pubkey,
+        expected_mint: &Pubkey,
+        token_program: &Pubkey,
+    ) -> Result<Self> {
+        let refund_account_info = remaining_accounts
+            .get(index)
+            .ok_or(ErrorCode::AccountNotEnoughKeys)?;
+
+        Self::try_token_account_from_account_info(
+            refund_account_info,
+            expected_token_account,
+            expected_mint,
+            token_program,
+        )
+        .map(Self::TokenAccount)
+        .or_else(|| {
+            Self::try_claim_account_from_account_info(
+                refund_account_info,
+                expected_mint,
+                expected_token_account,
+            )
+            .map(Self::ClaimAccount)
+        })
+        .ok_or_else(|| {
+            error::Error::from(CustomError::InvalidRefund)
+                .with_account_name(&format!("remaining_accounts[{}]", index))
+        })
+    }
+
+    // This implements the following Anchor account constraints when parsing remaining account as a token account:
+    // #[account(
+    //     mut,
+    //     address = expected_token_account @ CustomError::InvalidRefund,
+    //     token::mint = expected_mint,
+    //     token::token_program = token_program
+    // )]
+    // pub token_account: InterfaceAccount<'info, TokenAccount>,
+    // Note: All errors are ignored and Option is returned as we do not log them anyway due to memory constraints.
+    fn try_token_account_from_account_info(
+        account_info: &'info AccountInfo<'info>,
+        expected_token_account: &Pubkey,
+        expected_mint: &Pubkey,
+        token_program: &Pubkey,
+    ) -> Option<InterfaceAccount<'info, TokenAccount>> {
+        // Checks ownership on deserialization for the TokenAccount interface.
+        let token_account: InterfaceAccount<'info, TokenAccount> =
+            InterfaceAccount::try_from(account_info).ok()?;
+
+        // Checks if the token account is writable.
+        if !account_info.is_writable {
+            return None;
+        }
+
+        // Checks the token address matches.
+        if account_info.key != expected_token_account {
+            return None;
+        }
+
+        // Checks if the token account is associated with the expected mint.
+        if &token_account.mint != expected_mint {
```

Comment:
this is a meta-comment but this check has a `&` (borrowing) and the others don't. `expected_token_account` is passed in `&` but is not checked against the account_info with borrowing. 

point is: is there away to do this more consistently in this file and beyond?

Reply:
but you're right that its not fully consistent with comparison in `try_claim_account_from_account_info` below, so I just pushed commit to use `account_info.key != &pda_address` instead of `account_info.key() != pda_address` on line 121.